cumulus/parachains/common/src/storage_weight_reclaim.rs

```diff
@@ -0,0 +1,517 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Mechanism to reclaim PoV weight after an extrinsic has been applied.
+
+use codec::{Decode, Encode};
+use cumulus_primitives_core::Weight;
+use cumulus_primitives_proof_size_hostfunction::storage_proof_size::storage_proof_size;
+use frame_support::{
+	dispatch::{DispatchClass, DispatchInfo, PostDispatchInfo},
+	weights::WeightMeter,
+};
+use frame_system::{BlockWeight, Config};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, Dispatchable, PostDispatchInfoOf, SignedExtension},
+	transaction_validity::TransactionValidityError,
+	DispatchResult,
+};
+use sp_std::marker::PhantomData;
+
+const LOG_TARGET: &'static str = "runtime::storage_reclaim";
+
+/// Indicates that proof recording was disabled on the node side.
+const PROOF_RECORDING_DISABLED: u64 = u64::MAX;
+
+/// StorageWeightReclaimer is a mechanism for manually reclaiming storage weight.
+///
+/// It internally keeps track of the proof size and storage weight at initialization time. At
+/// reclaim  it computes the real consumed storage weight and refunds excess weight.
+///
+///
+/// # Example
+///
+/// ```ignore
+/// use parachains_common::storage_weight_reclaim::StorageWeightReclaimer;
+///
+/// // Initialize a StorageWeightReclaimer instance.
+/// let mut storage_reclaim = StorageWeightReclaimer::<Runtime>::start();
+///
+/// // Perform some operations that consume storage weight.
+///
+/// // Reclaim unused storage weight.
+/// if let Some(reclaimed_weight) = storage_reclaim.reclaim() {
+///     log::info!("Reclaimed {} weight", reclaimed_weight);
+/// }
+/// ```
+pub struct StorageWeightReclaimer<T: Config> {
+	previous_proof_weight: u64,
+	previous_reported_proof_size: Option<u64>,
```

Comment:
Nit: is it on purpose that one of the variables called `proof_weight`, and another `proof_size`? I can see it as convention that `weight` refers to benchmarked "weight" and `size` refers to actually consumed proof size, but below in `post_dispatch()` there is a variable called `consumed_weight`.

Reply:
I tried to use proof_size as the absolute number that is reported by the host function. Weight as the amount consumed or reported by the runtime facilities.